polynote-kernel/src/main/scala/polynote/kernel/util/TaskManager.scala

```diff
@@ -1,14 +1,160 @@
+                  semaphore.release
+
+          runTask.fiber.join.uncancelable.flatMap {
+            result =>
+              queued.onComplete(result, finalizer)
+          }.attempt.flatMap {
+            result =>
+              queued.deferred.complete(result) *> {
+                result match {
+                  case Left(err) => finalizer
+                  case Right(_)  => IO.unit
+                }
+              }
+          }
+      }
+  }
+
+  private val runQueueFiber: Ref[IO, Fiber[IO, Unit]] = Ref.unsafe(runNext().start.unsafeRunSync())
+
+  /**
+    * Cancel all queued tasks
+    */
+  def cancelAllQueued: IO[Unit] = IO {
+    val javaList = new java.util.ArrayList[Queued[_]](queuedTasks.size())
+    queuedTasks.drainTo(javaList)
+    javaList.asScala.toList
+  }.flatMap {
+    dequeuedTasks => dequeuedTasks.map {
+      queued => publish(queued.taskInfo.completed)
+    }.sequence.as(())
+  }
+
+  /**
+    * Queue a task which will result in a Stream. The next queued task will not start until the resulting stream
+    * terminates.
+    */
+  def queueTaskStream[A](id: String, label: String, detail: String = "")(fn: TaskInfo => IO[Stream[IO, A]]): IO[IO[Stream[IO, A]]] = {
+    val taskInfo = TaskInfo(id, label, detail, TaskStatus.Queued)
+    publish(taskInfo) *> Deferred[IO, Either[Throwable, Stream[IO, A]]].flatMap {
+      deferred => IO(queuedTasks.put(QueuedStream(taskInfo, fn(taskInfo), deferred))).map {
+        _ => deferred.get.flatMap {
+          case Left(err) => IO.raiseError(err)
+          case Right(result) => IO.pure(result)
+        }
+      }
+    }
+  }
+
+  /**
+    * Run a task by passing a function which will receive the [[TaskInfo]] and return a value. This bypasses the
+    * task queue.
+    *
+    * @return an IO, which will upon evaluation will "block" until the task runs and completes.
+    */
+  def runTask[A](id: String, label: String, detail: String = "")(fn: TaskInfo => A): IO[A] =
+    runTaskIO(id, label, detail)((taskInfo: TaskInfo) => IO.pure(taskInfo).map(fn))
+
+  /**
+    * Run a task by passing a function which will receive the [[TaskInfo]] and return an IO suspension which itself
+    * evaluates to a value. This bypasses the task queue.
+    *
+    * @return an IO, which will upon evaluation will "block" until the task runs and completes.
+    */
+  def runTaskIO[A](id: String, label: String, detail: String = "")(fn: TaskInfo => IO[A]): IO[A] = {
+    val taskInfo = TaskInfo(id, label, detail, TaskStatus.Queued)
+    statusUpdates.publish1(UpdatedTasks(taskInfo :: Nil)).unsafeRunAsyncAndForget()
+
+    val eval = for {
+      _       <- IO.cancelBoundary
+      runInfo  = taskInfo.copy(status = TaskStatus.Running)
+      runTask <- run(runInfo, fn(runInfo))
+      result  <- runTask.fiber.join.uncancelable.guarantee(IO(running.remove(runTask)))
+    } yield result
+
+    eval.guarantee {
+      publish(taskInfo.completed)
+    }
```

Comment:
Just wondering (no need to make this change if it works), but would it be wrong to move this guarantee up into the other guarantee above? Just thinking whether the two guarantees could be combined. 

Something like:

```
    for {
      _       <- IO.cancelBoundary
      runInfo  = taskInfo.copy(status = TaskStatus.Running)
      runTask <- run(runInfo, fn(runInfo))
      result  <- runTask.fiber.join.uncancelable.guarantee(IO(running.remove(runTask)) *> publish(taskInfo.completed))
    } yield result
```

Reply:
The upper guarantee only applies to the `fiber.join`, so it's not exactly equivalent... but maybe they could both be in the `eval.guarantee`.

Reply:
Ah, I see that makes sense - if something failed before getting to the `fiber.join` we'd never run those guarantees. And I guess if everything succeeded we would run the guarantees slightly earlier than we wanted to.